File: content/problems/1711-count-good-meals/solutions/power-complement-frequency/Solution.java
class Solution {
    public int countPairs(int[] deliciousness){
        final int MOD=1_000_000_007;
        Map<Integer,Integer> frequency=new HashMap<>();
        long answer=0;
        for(int value:deliciousness){
            for(int power=1;power<=1<<21;power<<=1)answer+=frequency.getOrDefault(power-value,0);
            answer%=MOD;
            frequency.put(value,frequency.getOrDefault(value,0)+1);
        }
        return (int)answer;
    }
}

File: content/problems/1711-count-good-meals/solutions/power-complement-frequency/Solution.kt
class Solution {
    fun countPairs(deliciousness:IntArray):Int{
        val mod=1_000_000_007L;
        val frequency=mutableMapOf<Int,Int>();
        var answer=0L;
        for(value in deliciousness){
            var power=1;
            while(power<=1 shl 21){
                answer+=frequency[power-value]?:0;
                power=power shl 1
            };
            answer%=mod;
            frequency[value]=(frequency[value]?:0)+1
        };
        return answer.toInt()
    }
}
